# <center> Lecture10 : Evaluating Regression Models </center>  
 
## <center> Instructor: Dr. Hu Chuan-Peng </center> 

* 在lec9中，我们已经学习了使用一个简单的线性模型来拟合变量间的关系。但这并未结束，接下来，我们需要对模型进行**评估**

* 我们可以从以下三个角度来思考这个问题：

    1. 这个模型有多公正？(How fair is the model?)

    2. 这个模型有多错误？(How wrong is the model?)

    3. 后验预测模型有多准确？(How accurate are the posterior predictive models?)


## 这个模型公正吗(Is the model fair?)

可以帮助我们进一步思考的几个问题：

1. 数据的收集过程是怎样的？

2. 数据由谁收集，数据收集的目的是什么？

3. 数据收集的过程，以及分析的结果，将对个人和社会产生什么影响？

4. 分析过程中可能会包含哪些偏见
    * p-hacking/HARKing

![Image Name](https://pic2.zhimg.com/80/v2-778de6c621356bc2c23c3a09ff2b0be5_1440w.webp)
(来源：胡传鹏, 王非, 过继成思, 宋梦迪, 隋洁, & 彭凯平. (2016). 心理学研究中的可重复性问题:从危机到契机. 心理科学进展(9), 15.)

## 这个模型可能有多错误(How wrong is the model?)

<center>“ all models are wrong, but some are useful.   ————George Box ”</center> 

* 尽管统计模型是对更复杂现实的简化表达，良好的统计模型仍然可以是有用的，并可以增进我们对世界复杂性的理解。

* 因此，在评估模型时，要问的下一个问题不是模型是否错误(is the model wrong?)，而是模型错误的程度(How wrong is the model?)

* 具体而言，在lec9中，贝叶斯线性回归模型的假设在多大程度上与现实相符？

### 重新审视模型假设

**Normal data model**  

在lec9中，我们使用一个线性模型来定义压力分数与自我控制分数之间的关系，并且指定了该模型成立的一些前提假设“

$$  
\begin{align*}  
Y_i &= \beta_0 + \beta_1 X_i + \epsilon \;\;\;\;\;\;\;\;\epsilon \sim N(0,\sigma^2)\\  
&\Downarrow \\  
Y_i | \beta_0, \beta_1, \sigma &\stackrel{ind}{\sim} N\left(\mu_i, \sigma^2\right) \;\; \text{ with } \;\; \mu_i = \beta_0 + \beta_1X_i  .\\  
\end{align*}  
$$  

* 回归模型需满足如下假设：  

    1. 独立观测假设:每个观测值$Y_i$是相互独立的，即一个观测的值不受其他观测的影响  

    2. 线性关系假设: 预测值$\mu_i$和自变量$X_i$之间可以用线性关系来描述，即：$\mu_i = \beta_0 + \beta_1 X_i$  

    3. 方差同质性假设： 在任意自变量的取值下，观测值$Y_i$都会以$\mu_i$为中心，同样的标准差$\sigma$呈正态分布变化（$\sigma$ is consistent）  

![Image Name](https://cdn.kesci.com/upload/s3y4rmsym1.png?imageView2/0/w/960/h/960)



**对模型的修改**

* 但并不是所有数据都会满足这些假设，当这些假设被违反时，我们需要考虑模型的修改：

**1. 当假设1被违反时：**

* 在心理学的实验数据当中，很多时候观测值之间存在依赖关系(dependent)。比如，反应时在单个被试内、或某种特定刺激类型内可能更同质：有的被试总是反应得比其他人更快，有的刺激总是能导致更快的反应。

* 这时，被试内/项目内的反应时数据是相互关联的

* 我们在之后会学习使用层级贝叶斯模型来处理相互关联的数据。

![Image Name](https://cdn.kesci.com/upload/s3xt2u50rw.png?imageView2/0/w/960/h/960)


**2. 当假设2和假设3被违反时：**

* 在左图中，我们可以看到，Y和X之间的关系并非线性的

* 并且，随着X的增大，Y的变异性越来越大。

* 在右图中可以看到，后验预测分布有实际观测值的分布差异很大

![Image Name](https://www.bayesrulesbook.com/bookdown_files/figure-html/nonlinear-1-ch10-1.png)

**有两种方式来处理这个问题**

**a. 使用不同的数据模型**

* 不假设实际值与观测值之间的关系是正态的  $Y_i \sim N(\mu_i, \sigma^2)$

* 之后，我们会学习使用其他回归模型来描述其数据关系，比如泊松回归、二项回归、负二项回归。

**b. 对数据进行变换**

> 如果数据模型并不是我们要担心的问题，我们可以对数据进行变换，仍然可以对变换后的数据可以使用正态模型：

* 对$Y$进行变换：$g(Y_i) | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2)$，$\mu_i = \beta_0 + \beta_1 X_i$

* 对$X$进行变换： $Y_i | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2)$，$\mu_i = \beta_0 + \beta_1 h(X_i)$

* 同时对$Y$和$X$进行变换：$g(Y_i) | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2)$， $\mu_i = \beta_0 + \beta_1 h(X_i)$

在刚刚的例子当中，我们对$Y$的取值做一个对数变换

$$\log(Y_i) | \beta_0, \beta_1, \sigma \stackrel{ind}{\sim}N(\mu_i, \sigma^2) \;\; \text{ with } \; \mu_i = \beta_0 + \beta_1 X_i$$

* 在变换之后，可以看到$\log(Y)$与$X$之间的关系仍然是线性的，且随着$X$的增大，$Y$的变异性仍然是一致的。

* 可以使用正态的线性模型来拟合$\log(Y)$的后验分布

    ![Image Name](https://www.bayesrulesbook.com/bookdown_files/figure-html/nonlinear-2-ch10-1.png)

## 模型评估与比较

* 以上是模型选定之前需要考虑的一些问题，当我们确定选用某几个模型之后，我们需要对模型的表现进行评估与比较

* 常用的模型评估与比较的统计指标可被分为三类：

    1. 模型的拟合优度(Goodness of fit)

    2. 对新数据的预测准确性

    3. 边缘似然理论的指标


### 拟合优度(Goodness of fit)  

* 拟合优度衡量模型在**样本集**上拟合的程度，并没有考虑模型的复杂度，只是单纯地衡量模型对当前数据的拟合的好坏。  

    * 比如回归模型中的$R^2$，即回归模型所能解释的因变量的变异的百分比

* 此外，这里我们还介绍了其他指标：

    1. 对数似然(log likelihood function) 

    2. within_94

    3. 均方差(Mean squared error, MSE)
    
    

资料来源：https://baike.baidu.com/pic/%E6%8B%9F%E5%90%88%E4%BC%98%E5%BA%A6/10946257/0/6a63f6246b600c338b38d998114c510fd9f9a118

> 我们先运行模型定义、MCMC采样的代码

In [127]:
# 导入 pymc 模型包，和 arviz 等分析工具 
import pymc as pm
import arviz as az
import seaborn as sns
import scipy.stats as st
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr
import pandas as pd

# 忽略不必要的警告
import warnings
warnings.filterwarnings("ignore")

In [128]:
# 通过 pd.read_csv 加载数据 Data_Sum_HPP_Multi_Site_Share.csv
df_raw = pd.read_csv('Data_Sum_HPP_Multi_Site_Share.csv')

# 筛选站点为"Tsinghua"的数据
df = df_raw[df_raw["Site"] == "Tsinghua"]

# 选择需要的两列
df = df[["stress","scontrol"]]

#设置索引
df["index"] = range(len(df))
df = df.set_index("index")


In [129]:
with pm.Model() as linear_model:

    beta_0 = pm.Normal("beta_0", mu=70, sigma=4)          #定义beta_0          
    beta_1 = pm.Normal("beta_1", mu=0, sigma=0.8)         #定义beta_1
    sigma = pm.Exponential("sigma", 0.6)                  #定义sigma

    x = pm.MutableData("x",df.stress)                     #x是自变量压力水平

    mu = beta_0 + beta_1 * x                              #定义mu，讲自变量与先验结合

    likelihood = pm.Normal("y_est", mu=mu, sigma=sigma, observed=df.scontrol)   #定义似然：预测值y符合N(mu, sigma)分布
                                                                                #通过 observed 传入实际数据y 自我控制水平

In [141]:
#===========================
#     注意！！！以下代码可能需要运行1-2分钟左右
#===========================
with linear_model:
    trace = pm.sample(draws=5000,                   # 使用mcmc方法进行采样，draws为采样次数
                      tune=1000,                    # tune为调整采样策略的次数，可以决定这些结果是否要被保留
                      chains=4,                     # 链数
                      discard_tuned_samples= True,  # tune的结果将在采样结束后被丢弃
                      random_seed=84735)

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_1, sigma]


Sampling 4 chains for 1_000 tune and 5_000 draw iterations (4_000 + 20_000 draws total) took 103 seconds.


#### 对数似然

* 对于每一个后验预测值，计算它在当前模型下发生的可能性乘积

$$f\left(y' | \vec{y}\right) = \int\int\int f\left(y' | \beta_0,\beta_1,\sigma\right) f(\beta_0,\beta_1,\sigma|y) d\beta_0 d\beta_1 d\sigma$$

* $f\left(y' | \beta_0,\beta_1,\sigma\right)$ 表示$y'$在当前模型下出现的可能性

* $f(\beta_0,\beta_1,\sigma|y)$ 表示参数的后验概率


* 由于乘积的形式会得到一个很小的数，不便于后续处理，因此将似然取对数以便比较。 
    * 如果模型的预测越准确，则后验预测值出现的可能性越大

$$log(f\left(y' | \vec{y}\right))=\sum log(f\left(y' | \vec{y}\right))$$

![Image Name](https://www.bayesrulesbook.com/bookdown_files/figure-html/post-example-ch9-1.png)

* 在pymc中，我们可以使用`pm.compute_log_likelihood`和`az.loo`来直接计算对数似然

* 在`az.loo`返回的值中，`elpd_loo`表示$\log(f(y_{\text{new}} | \vec{y}))$的pdf均值，`elpd_loo`越高表示模型的预测值越精确

In [ ]:
with linear_model:
    pm.compute_log_likelihood(trace)

In [147]:
az.loo(trace)

Computed from 20000 posterior samples and 174 observations log-likelihood matrix.

         Estimate       SE
elpd_loo  -545.76     9.54
p_loo        2.72        -
------

Pareto k diagnostic values:
                         Count   Pct.
(-Inf, 0.5]   (good)      174  100.0%
 (0.5, 0.7]   (ok)          0    0.0%
   (0.7, 1]   (bad)         0    0.0%
   (1, Inf)   (very bad)    0    0.0%

>`elpd_loo`的值为-545.76 ，但仅凭单个值，我们并没有一个比较的参照，来告诉我们模型的预测究竟精确到了一个什么程度。因此，这个指标在模型比较中的用处更大。

#### within_94

$$
\begin{array}{lll}
\text{Markov chain parameter sets} & & \text{Simulated posterior predictive models} \\
\left[
\begin{array}{lll} 
\beta_0^{(1)} & \beta_1^{(1)} & \sigma^{(1)} \\
\hline
\beta_0^{(2)} & \beta_1^{(2)} & \sigma^{(2)} \\
\hline
\vdots & \vdots & \vdots \\
\hline
\beta_0^{(20000)} & \beta_1^{(20000)} & \sigma^{(20000)} \\
\end{array}
\right]
& \; \longrightarrow \;
& \left[
\begin{array}{l|l|l|l} 
Y_1^{(1)} & Y_2^{(1)} & \cdots & Y_{174}^{(1)} \\
Y_1^{(2)} & Y_2^{(2)} & \cdots & Y_{174}^{(2)} \\
\vdots & \vdots &  & \vdots \\
Y_1^{(20000)} & Y_2^{(20000)} & \cdots & Y_{174}^{(20000)} \\
\end{array}
\right]
\end{array}

$$

* 在当前数据集中，自变量的取值共有`174`个

* 代入`20000`个参数对，每个自变量都拥有对应的20000个后验预测值

* 对于这些后验预测值，我们可以计算其可信区间，并观察真实值落在可信区间的情况

* `az.summary`直接给出了每个预测值94%的可信区间

*（注：可信区间范围可以改变）*

In [ ]:
with linear_model:
    trace = pm.sample_posterior_predictive(trace)

Sampling: [y_est]


In [132]:
az.summary(trace)

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
y_est[0]    40.487  5.538  30.229   50.949      0.039    0.027   20387.0   
y_est[1]    37.999  5.506  27.670   48.349      0.040    0.028   18871.0   
y_est[2]    44.066  5.534  33.530   54.404      0.040    0.028   19205.0   
y_est[3]    42.049  5.560  31.596   52.487      0.041    0.029   18603.0   
y_est[4]    36.967  5.558  26.537   47.334      0.040    0.028   19621.0   
...            ...    ...     ...      ...        ...      ...       ...   
y_est[169]  43.571  5.591  33.366   54.458      0.040    0.028   19786.0   
y_est[170]  38.515  5.532  27.745   48.642      0.039    0.028   19676.0   
y_est[171]  38.418  5.545  28.031   48.807      0.040    0.028   19628.0   
y_est[172]  40.050  5.555  29.568   50.467      0.039    0.028   19987.0   
y_est[173]  43.010  5.565  32.150   53.078      0.039    0.028   19923.0   

            ess_tail  r_hat  
y_est[0]     19853.0    1.0  
y_est[1]     19060.0    1.0  
y_est[2]     19496.0    1.0  
y_est[3]     19359.0    1.0  
y_est[4]     19424.0    1.0  
...              ...    ...  
y_est[169]   18562.0    1.0  
y_est[170]   19764.0    1.0  
y_est[171]   19664.0    1.0  
y_est[172]   19856.0    1.0  
y_est[173]   19561.0    1.0  

[174 rows x 9 columns]

**within_94可视化**

![Image Name](https://cdn.kesci.com/upload/s3xt34m35o.png?imageView2/0/w/960/h/960)


#### 均方差(Mean squared error, MSE)

* 计算公式为：$MSE(\hat{\theta})=E(\hat{y}-y)^2=\frac{1}{n}\sum (\hat{y}-y)^2$

* 均方差表示预测变量与目标变量之间差值的平方之和的均值

![Image Name](https://cdn-media-1.freecodecamp.org/images/MNskFmGPKuQfMLdmpkT-X7-8w2cJXulP3683)

资料来源：https://www.freecodecamp.org/news/machine-learning-mean-squared-error-regression-line-c7dde9a26b93/

我们可以在`az.summary(trace)`生成的表格基础上，总结均方差与`within_94`

In [ ]:
az.summary(trace)

In [137]:
# 将az.summary(trace)的结果存入pred_sum_table 中
pred_sum_table = az.summary(trace)

# 重新设置索引
pred_sum_table = pred_sum_table.reset_index()

#将真实数据集中的自变量与因变量合并到当前表格
pred_sum_table["x_obs"] = trace.constant_data["x"].values
pred_sum_table["y_obs"] = trace.observed_data["y_est"].values

#代入均方差公式，计算均方差
pred_sum_table['MSE'] = pred_sum_table['y_obs'] - pred_sum_table['mean']
pred_sum_table['MSE'] **= 2

#生成新的一列，判断每个真实观测值y是否在后验预测值94%的可信区间内，若是则记为1，否则记为0
pred_sum_table['verify'] = np.where((pred_sum_table['y_obs'] > pred_sum_table['hdi_3%']) & (pred_sum_table['y_obs'] < pred_sum_table['hdi_97%']), 1, 0)

In [138]:
pred_sum_table

index    mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
0      y_est[0]  40.487  5.538  30.229   50.949      0.039    0.027   20387.0   
1      y_est[1]  37.999  5.506  27.670   48.349      0.040    0.028   18871.0   
2      y_est[2]  44.066  5.534  33.530   54.404      0.040    0.028   19205.0   
3      y_est[3]  42.049  5.560  31.596   52.487      0.041    0.029   18603.0   
4      y_est[4]  36.967  5.558  26.537   47.334      0.040    0.028   19621.0   
..          ...     ...    ...     ...      ...        ...      ...       ...   
169  y_est[169]  43.571  5.591  33.366   54.458      0.040    0.028   19786.0   
170  y_est[170]  38.515  5.532  27.745   48.642      0.039    0.028   19676.0   
171  y_est[171]  38.418  5.545  28.031   48.807      0.040    0.028   19628.0   
172  y_est[172]  40.050  5.555  29.568   50.467      0.039    0.028   19987.0   
173  y_est[173]  43.010  5.565  32.150   53.078      0.039    0.028   19923.0   

     ess_tail  r_hat  x_obs  y_obs         MSE  verify  
0     19853.0    1.0     38   47.0   42.419169       1  
1     19060.0    1.0     43   40.0    4.004001       1  
2     19496.0    1.0     31   40.0   16.532356       1  
3     19359.0    1.0     35   46.0   15.610401       1  
4     19424.0    1.0     45   50.0  169.859089       0  
..        ...    ...    ...    ...         ...     ...  
169   18562.0    1.0     32   52.0   71.048041       1  
170   19764.0    1.0     42   33.0   30.415225       1  
171   19664.0    1.0     42   35.0   11.682724       1  
172   19856.0    1.0     39   38.0    4.202500       1  
173   19561.0    1.0     33   39.0   16.080100       1  

[174 rows x 14 columns]

In [139]:
#将均方差，与within_94的平均值放入一个新表格中
prediction_summary = pd.DataFrame({"MSE_mean": [pred_sum_table['MSE'].sum()/len(pred_sum_table['MSE'])],
                                   "within_94%": [pred_sum_table['verify'].mean()]})

prediction_summary

MSE_mean  within_94%
0  30.057599    0.931034

### 对新数据的预测准确性